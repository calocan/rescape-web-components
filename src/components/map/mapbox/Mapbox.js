/**
 * Created by Andy Likuski on 2017.02.16
 * Copyright (c) 2017 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import reactMapGl from 'react-map-gl';
import {reqPathThrowing, reqStrPathThrowing} from 'rescape-ramda';
import {
  composeViews, renderChoicepoint, propsFor,
  propsForSansClass, renderErrorDefault, renderLoadingDefault
} from 'rescape-helpers-component';
import {eMap, nameLookup} from 'rescape-helpers-component'
import * as R from 'ramda';
import {styleMultiplier} from 'rescape-helpers-component';
import {applyMatchingStyles, mergeAndApplyMatchingStyles} from 'rescape-apollo';
import {Component} from 'react'
import PropTypes from 'prop-types';


const [Div, ReactMapGl] = eMap(['div', reactMapGl]);
export const c = nameLookup({
  mapbox: true,
  mapboxMapReactGlOuter: true,
  mapboxReactMapGl: true,
  mapboxLoading: true,
  mapboxError: true
});

/**
 * The View for Mapbox
 */
class Mapbox extends Component {
  render() {
    const props = Mapbox.views(this.props);
    return Div(propsFor(props.views, c.mapbox),
        Mapbox.choicepoint(props)
    );
  }
}

Mapbox.viewStyles = ({style}) => {
  return {
    [c.mapbox]: mergeAndApplyMatchingStyles(style, {
      position: 'absolute',
      width: styleMultiplier(1),
      height: styleMultiplier(1)
    }),

    [c.mapboxReactMapGl]: applyMatchingStyles(style, {
      width: styleMultiplier(1),
      height: styleMultiplier(1)
    }, [c.sankeyFiltererItem])
  };
};

Mapbox.viewProps = (props) => {
  return {
    [c.mapboxReactMapGl]: R.mergeAll([
      {
        // Width and height are calculated in viewStyles
        width: reqStrPathThrowing(`views.${c.mapboxReactMapGl}.style.width`),
        height: reqStrPathThrowing(`views.${c.mapboxReactMapGl}.style.height`)
      },
      // Pass anything in mapbox
      reqStrPathThrowing('data.mapbox', props),
      // Pass anything in viewport
      reqStrPathThrowing('data.viewport', props)
    ])
    //osm: 'region.geojson.osm'
  };
};

Mapbox.viewActions = () => {
  return {
    [c.mapboxReactMapGl]: ['onViewportChange', 'hoverMarker', 'selectMarker']
  };
};

Mapbox.renderData = ({views}) => {
  /* We additionally give Mapbox the container width and height so the map can track changes to these
   We have to apply the width and height fractions of this container to them.
   */
  const props = propsFor(views);
  const propsSansClass = propsForSansClass(views);

  return Div(props(c.mapboxMapReactGlOuter),
    ReactMapGl(propsSansClass(c.mapboxReactMapGl))
  );
};

/**
 * Adds to props.views for each component configured in viewActions, viewProps, and viewStyles
 * @param {Object} props this.props or equivalent for testing
 * @returns {Object} modified props
 */
Mapbox.views = composeViews(
  Mapbox.viewActions(),
  Mapbox.viewProps,
  Mapbox.viewStyles
);

/**
 * Loading, Error, or Data based on the props
 */
Mapbox.choicepoint = renderChoicepoint(
  renderErrorDefault(c.mapboxError),
  renderLoadingDefault(c.mapboxLoading),
  Mapbox.renderData
);

Mapbox.propTypes = {
  data: PropTypes.shape().isRequired,
  style: PropTypes.shape().isRequired
};

export default Mapbox;
