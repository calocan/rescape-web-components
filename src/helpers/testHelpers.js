/**
 * Created by Andy Likuski on 2018.01.17
 * Copyright (c) 2018 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {getCurrentConfig, privateConfig, createSchema, createInitialState} from 'rescape-sample-data';
import {createSelectorResolvedSchema} from 'rescape-apollo';
import {of} from 'folktale/concurrency/task';
import PropTypes from 'prop-types';
import {v} from 'rescape-validate';
import {makeSampleInitialState, makeApolloTestPropsTaskFunction, propsFromParentPropsTask} from 'rescape-helpers-test'


/**
 *
 * These helpers are trivial functions that use rescape-sample-data and rescape-helpers-component togther,
 * creating functions that are seeded with sample data for testing
 */
export const currentConfig = getCurrentConfig(privateConfig);

/**
 * Sample initial state for tests
 */
export const sampleInitialState = makeSampleInitialState(createInitialState, currentConfig);

/**
 * Resolved schema for tests
 * @type {Object}
 */
export const resolvedSchema = createSelectorResolvedSchema(createSchema(), currentConfig);

/**
 * Calls makenApolloTestPropsFunction with the given * 3 arguments:
 * mapStateToProps, mapDispatchToProps, {query, args}
 * See makeApolloTestPropsFunction for details.
 * The resulting promise is wrapped in a Task.
 * @param {Function} mapStateToProps
 * @param {Function} mapDispatchToProps
 * @param {Object} queryInfo
 * @returns {Function} A 2 arity function called with state and props that results in a Task that
 * resolves the props as A Right if no errors and a Left if errors
 */
export const apolloTestPropsTaskMaker = v((mapStateToProps, mapDispatchToProps, queryInfo) =>
    (state, props) => makeApolloTestPropsTaskFunction(
      resolvedSchema,
      currentConfig,
      mapStateToProps,
      mapDispatchToProps,
      queryInfo
    )(state, props)
  , [
    ['mapStateToProps', PropTypes.func.isRequired],
    ['mapDispatchToProps', PropTypes.func.isRequired],
    ['queryInfo', PropTypes.shape().isRequired]
  ], 'apolloTestPropsTaskMaker');


/**
 * Given a Task to fetch parent container props and a task to fetch the current container props,
 * Fetches the parent props and then samplePropsTaskMaker with the initial state and parent props
 * @param {Task} parentContainerSamplePropsTask Task that resolves to the parent container props
 * @param {Function} samplePropsTaskMaker 2 arity function expecting state and parent props.
 * Returns a Task from a container that expects a sample state and sampleOwnProps
 * and then applies the container's mapStateToProps, mapDispatchToProps, and optional mergeProps
 * @param parentComponentViews A function expecting props that returns an object keyed by view names
 * and valued by view props, where views are the child containers/components of the component
 * @param viewName The viewName in the parent component of the target container
 * @returns {Task} A Task to asynchronously return the parentContainer props merged with sampleOwnProps
 * in an Result.Ok. If anything goes wrong an Result.Error is returned
 */
export const localPropsFromParentPropsHelperTask = v((chainedParentPropsTask, samplePropsTaskMaker) =>
  propsFromParentPropsTask(sampleInitialState, chainedParentPropsTask, samplePropsTaskMaker),
  [
    ['chainedParentPropsTask', PropTypes.shape().isRequired],
    ['samplePropsTaskMaker', PropTypes.func.isRequired]
  ],
  'localPropsFromParentPropsHelperTask');

