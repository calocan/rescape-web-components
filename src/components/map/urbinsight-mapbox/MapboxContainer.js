/**
 * Created by Andy Likuski on 2017.02.06
 * Copyright (c) 2017 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {onViewportChange} from 'redux-map-gl';
import {makeMergeDefaultStyleWithProps} from 'rescape-apollo';
import {mapboxSelector, viewportSelector} from 'rescape-apollo';
import {makeActiveUserAndSettingsSelector} from 'rescape-apollo';
import {createSelector} from 'reselect';
import {makeApolloTestPropsFunction} from 'rescape-helpers-component';
import {mergeDeep} from 'rescape-ramda';
import Mapbox from './Mapbox';
import * as R from 'ramda';
import {v} from 'rescape-validate';
import PropTypes from 'prop-types';
import {composeGraphqlQueryDefinitions} from 'rescape-helpers-component';


/**
 * Selects the current user from state
 * and the Viewport for the region in the props
 * @params {Object} state
 * @params {Object} props
 * @params {Object} props.user User values
 * @params {Object} props.scope Region, project, scenario, etc. defining the scope.
 * @params {Object} props.scope.region A region containing at least an id. If Mapbox doesn't need the region pass
 * the global region
 * @params {Object} [props.scope.project] A project containing at least an id. If Mapbox doesn't need the region pass
 * the global region
 * @params {Object} [props.scope.scenario] A scenario containing at least an id.
 * @returns {Object} The selected props
 */
export const mapStateToProps = v((state, props) => {
    const {style, ...data} = props;
    return createSelector(
      [
        makeActiveUserAndSettingsSelector(),
        makeMergeDefaultStyleWithProps(),
        mapboxSelector
      ],
      (userAndSettings, defaultStyle, {viewport, ...mapbox}) => ({
        data: R.mergeAll([
          userAndSettings,
          // Mapbox is selected separately to combine region.mapbox with settings.mapbox
          // Viewport is combined with other properties in the react-map-gl component, hence separated here
          {viewport, mapbox},
          data
        ]),
        style: R.merge(defaultStyle, style)
      })
    )(state, props);
  },
  [
    ['state', PropTypes.shape().isRequired],
    ['props', PropTypes.shape({
      scope: PropTypes.shape({
        region: PropTypes.shape().isRequired,
        project: PropTypes.shape(),
        scenario: PropTypes.shape(),
      }).isRequired
    }).isRequired]
  ], 'mapStateToProps');

export const mapDispatchToProps = (dispatch, ownProps) => {
  return bindActionCreators({
    onViewportChange
    //hoverMarker,
    //selectMarker
  }, dispatch);
};


/**
 * Query
 * Prerequisites:
 *   A Region
 * Resolves:
 *  The geojson of the Region
 * Without prerequisites:
 *  Skip render, although in the future we'll want to use this component to choose a region
 */
const geojsonQuery = `
    query geojson($regionId: String!) {
          region(id: $regionId) {
              id
              geojson {
                  osm {
                      features {
                          id
                          type
                          geometry {
                              type
                              coordinates
                          }
                          properties
                      }
                  }
              }
        }
    }
`;

// TODO this needs to take viewport args
const viewportMutation = `
  mutation mutateViewport($isConnected: Boolean) {
    mutateViewport(isConnected: $isConnected) @client
  }
`;

/**
 * All queries used by the container
 */
export const queries = {
  /**
   * Expects a region with an id and resolves geojson of the region
   */
  geojson: {
    query: geojsonQuery,
    args: {
      options: ({data: {region}}) => ({
        variables: {
          regionId: region.id
        },
        // Pass through error so we can handle it in the component
        errorPolicy: 'all'
      }),
      props: ({data, ownProps}) => mergeDeep(
        ownProps,
        {data}
      )
    }
  },
  // Mutate the Mapbox viewport, storing it in local cache
  mutateViewport: {
    query: viewportMutation,
    args: {
      options: {
        errorPolicy: 'all'
      },
      props: ({mutate}) => ({
        mutateViewport:
          (filterNodeCategory, filterNodeValue) => mutate({variables: {filterNodeCategory, filterNodeValue}})
      })
    }
  }
};

// Create the GraphQL Container.
const ContainerWithData = composeGraphqlQueryDefinitions(queries)(Mapbox);

// Using R.merge to ignore ownProps, which were already merged by mapStateToProps
export default connect(mapStateToProps, mapDispatchToProps, R.merge)(ContainerWithData);

