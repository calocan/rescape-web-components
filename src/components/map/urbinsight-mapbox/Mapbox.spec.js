import mapGL from 'react-map-gl';
import mapbox from './Mapbox';
import {shallow} from 'enzyme';
import {propsFromSampleStateAndContainer} from 'rescape-helpers-test';
import {defaultRunConfig} from 'rescape-ramda';
import {eMap} from 'rescape-helpers-component';
import {sampleInitialState} from '../../../helpers/testHelpers';
import {samplePropsTaskMaker} from './MapboxContainer.sample';

const [MapGL, Mapbox] = eMap([mapGL, mapbox]);

describe('Mapbox', () => {
  test('Can mount', done => {
    propsFromSampleStateAndContainer(sampleInitialState, samplePropsTaskMaker,
      {
        // style dimensions are normally from the parent
        style: {
          width: 500,
          height: 500
        }
      }
    ).run().listen(defaultRunConfig({
      onResolved:
        props => {
          const wrapper = shallow(
            Mapbox(props)
          );
          expect(wrapper).toMatchSnapshot();
        }
    }));
  });
});

