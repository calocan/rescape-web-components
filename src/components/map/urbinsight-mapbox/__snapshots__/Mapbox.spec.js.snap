// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mapbox Can mount 1`] = `
<div
  className="mapbox-root"
>
  <InteractiveMap
    altitude={1.5}
    attributionControl={true}
    bearing={0}
    clickRadius={0}
    doubleClickZoom={true}
    dragPan={true}
    dragRotate={true}
    getCursor={[Function]}
    iconAtlas="data/location-icon-atlas.png"
    mapStyle="mapbox://styles/mapbox/light-v8"
    mapboxApiAccessToken="pk.eyJ1IjoiY2Fsb2NhbiIsImEiOiJjaXl1aXkxZjkwMG15MndxbmkxMHczNG50In0.07Zu3XXYijL6GJMuxFtvQg"
    maxLatitude={85.05113}
    maxPitch={60}
    maxZoom={20}
    minLatitude={-85.05113}
    minPitch={0}
    minZoom={0}
    onClick={null}
    onHover={null}
    onLoad={[Function]}
    onViewportChange={null}
    perspectiveEnabled="true"
    pitch={0}
    preserveDrawingBuffer={false}
    preventStyleDiffing="false"
    scrollZoom={true}
    showCluster="true"
    showZoomControls="true"
    touchZoomRotate={true}
    viewport={
      Immutable.Map {
        "pitch": 40,
        "bearing": 0,
        "startDragLngLat": null,
        "isDragging": false,
        "latitude": 37,
        "longitude": -119,
        "zoom": 5,
      }
    }
    visibilityConstraints={
      Object {
        "maxLatitude": 85.05113,
        "maxPitch": 60,
        "maxZoom": 20,
        "minLatitude": -85.05113,
        "minPitch": 0,
        "minZoom": 0,
      }
    }
    visible={true}
  />
</div>
`;
